Feature gg multiline paste

gordonguthrie opened the pull request.

Commit:
interim commit of riak_shell changes for multiline paste

Commit:
Major reorganisation of the code paths plus bug fix.

The old cycle was that the prompt system would return a string command "show_tables;" which
would then be parsed into Tokens.

Now the prompt returns Tokens - but various things like logging, history require the string command

The string command is recreated from the tokens - this simplifies the API but requires a big
dance down all the function calls in the call cycle

There was an (unreported) where using 'show_history' as the first command would crash the shell.
This has been fixed.

Commit:
minor alignment typo

andytill commented on include/riak_shell.hrl:
```diff
@@ -10,11 +10,10 @@
-    partial_cmd            = [] :: [char()],
-    partial_tokens         = [] :: [term()],
-    response               = [] :: [char()]
+    cmd                    = []    :: [char()],
```
Could be `string()`.

andytill commented:
As discussed, single quotes are being stripped from queries e.g. `SELECT id FROM docstab WHERE id = 'lol' AND  time > 10 AND time < 1000;`

**Fixed in** 2332430

andytill commented:
There is a bug when a semi colon appears within single quotes.

```
riak-shell(1)>SELECT id FROM docstab WHERE id = 'l;ol' AND  time > 10 AND time < 1000;
SQL Parser error {<<"ol">>,riak_ql_parser,
                  ["syntax error before: ","identifier"]}
```

Works fine with double quotes.

**UPDATE 14/11/2016**

This is fixed.

``` sql
riak-shell(9)>SELECT a FROM docstab WHERE a = 'l;ol' AND  b > 10 AND b < 1000;        
riak-shell(10)>
```

andytill commented:
After executing a statement, press enter a few times and then execute another statement. The second statement will fail with `Invalid Command`.

```
riak-shell(1)>CREATE TABLE mtest4( id SINT64 NOT NULL, PRIMARY KEY((id),id));
riak-shell(1)>
riak-shell(1)>
riak-shell(1)>
riak-shell(2)>
riak-shell(2)>
riak-shell(2)>
riak-shell(2)>CREATE TABLE mtest5( id SINT64 NOT NULL, PRIMARY KEY((id),id));
Invalid Command:  CREATE TABLE mtest5( id SINT64 NOT NULL, PRIMARY KEY((id),id));
```

**Fixed in** 7ee5d53.

andytill commented:
If two commands are on the same line then the second command is ignored.

```
riak-shell(1)>CREATE TABLE mtest8( id SINT64 NOT NULL, PRIMARY KEY((id),id)); SHOW TABLES;
riak-shell(2)>
```

**UPDATE 14/11/2016** create a JIRA for this https://bashoeng.atlassian.net/browse/RTS-1524

Commit:
Ignore whitespace at the start of a statement.

Commit:
Do not strip single quotes from atoms, to allow varchars.

andytill commented:
This change doesn't allow coping in many statements and have them execute one after the other, for example it would be very useful in demo scenarios to run up the shell and paste a page of commands, in this case the descends keys docs I am writing.

``` sql
CREATE TABLE descending_table (
a SINT64 NOT NULL,
b TIMESTAMP NOT NULL,
PRIMARY KEY ((a, quantum(b, 1, 'm')), a, b DESC));

INSERT INTO descending_table VALUES (1,1);
INSERT INTO descending_table VALUES (1,2);
INSERT INTO descending_table VALUES (1,3);
INSERT INTO descending_table VALUES (1,4);
INSERT INTO descending_table VALUES (1,5);

SELECT * FROM descending_table WHERE a = 1 AND b >= 1 AND b <= 5;

+-+------------------------+
|a|           b            |
+-+------------------------+
|1|1970-01-01T00:00:00.005Z|
|1|1970-01-01T00:00:00.004Z|
|1|1970-01-01T00:00:00.003Z|
|1|1970-01-01T00:00:00.002Z|
|1|1970-01-01T00:00:00.001Z|
+-+------------------------+
```

**UPDATE 14/11/2016** create a JIRA for this https://bashoeng.atlassian.net/browse/RTS-1524

Commit:
interim commit - to squash

Commit:
Continued refactoring of riak_shell

gordonguthrie commented:
`riak-shell` now implements the erlang shell prompt/input now - that doesn't support multi-statement paste - that is an input string with multiple `$end` tokens - because it returns on `$end` so to do this would involve reimplementing a chunk of the io system libraries...

gordonguthrie commented:
I was going to squash these commits but the merge on the git log makes that a very problematic process.

gordonguthrie commented:
This would have been the squashed commit message

Major refactor of riak_shell to enable multiline pasting

The initial idea for riak-shell was for a learning exercise for developers moving over to use
leex/yecc - the Erlang lexer/parsers that are used in riak_ql.

In the initial version the prompt returned an input string which could then be lexed.
The lexer tokens were then inspected and, if appropriate, the input string was forwarded
to the riak_ql SQL lexer/parser - if not the riak-shell parser was used.

Switching to the proper Erlang input means that the prompt now returns a token string which must
be reversed to make the input. This made it necessary to turn the cmd_line from a poorly specified
'toy' language to a slightly-less toy one - but one that was better aligned to SQL in how it handled
edge cases.

A lot of the work in this commit is the consequence of 'un-toying' the lexer and all the tech
debt that came with it.

Commit:
Fixes consequent to running riak_shell regression tests

This fix:
* changes the way in which replay command results are checked
* cleans up and fixes the interface to riak_test
* stops about; crashing replay shells in riak_test

andytill commented:
Putting a new line in front of a command always results in a command line parser error like below. When repeated the command works fine.

```
riak-shell(2)>INSERT INTO keytab VALUES (2,1); 
riak-shell(3)>INSERT INTO keytab VALUES (3,1);
riak-shell(4)>
riak-shell(4)>INSERT INTO keytab VALUES (4,1);
Error: {error,{2,cmdline_parser,["syntax error before: ",["\"(\""]]}}
```

**Fixed in:** f6ab7dc

hazen commented:
@andytill @gordonguthrie Any word on this? I'm hoping to tweak `riak_shell`, but need this refactor in before even attempting it. Cheers!

andytill commented:
@javajolt @gordonguthrie no word, but this fixes a bunch of issues on it's own, so maybe we get this PR merged with the features it has today and JIRA up the other known issues?

Commit:
Fix for issue with leading whitespce in SQL commands causing problems for riak_shell

andytill commented:
Verified that this works, but an eunit is failing.

```
riak_shell: concat_atoms_with_underscore_test (module 'riak_shell')...*failed*
in function riak_shell:'-concat_atoms_with_underscore_test/0-fun-0-'/0 (src/riak_shell.erl, line 564)
in call from riak_shell:concat_atoms_with_underscore_test/0 (src/riak_shell.erl, line 563)
**error:{assertMatch_failed,[{module,riak_shell},
                     {line,564},
                     {expression,"make_riak_shell_cmd ( [ { atom , 1 , \"One\" } , { underscore , 1 , \"_\" } , { atom , 1 , \"Two\" } ] )"},
                     {pattern,"{ \"one_two\" , _ }"},
                     {value,{"one",[{underscore,1,"_"},{atom,1,"Two"}]}}]}


=======================================================
  Failed: 1.  Skipped: 0.  Passed: 0.
```

Commit:
The function that these tests tested has been refactored. We now concatenate inside
the lexer as this change set shows:

https://github.com/basho/riak_shell/blob/f6ab7dcc5feed945065e6bf06e233effcbe6d02e/src/cmdline_lexer.xrl#L27

Commit:
Exported a function for debugging - forgot to un-export it for the last commit

Commit:
remove errant io:formats

hazen commented on src/riak_shell.erl:
```diff
@@ -510,37 +495,63 @@ maybe_print_exception(#state{debug = on}, Type, Err) ->
--ifdef(TEST).
--include_lib("eunit/include/eunit.hrl").
-
-full_cmd_name_test() ->
```
Do these tests perhaps belong somewhere else?

Commit:
Merge remote-tracking branch 'origin/develop' into feature_gg_multiline_paste

Commit:
Remove failing steps from .thumbs.yml

Commit:
Use `make all` instead of `make compile`

andytill commented:
+1

hazen commented:
+1

korry8911 commented:
+1

davidx commented:
thumbot retry

davidx commented:
thumbot retry

davidx commented:
👍

davidpuddy1 commented:
+1

davidx commented:
+1
